File: android/app/src/main/java/com/nxtend/team35/yubiboard/vision/ArucoMarkerProcessor.kt
package com.nxtend.team35.yubiboard.vision

import android.os.SystemClock
import androidx.camera.core.ImageProxy
import com.nxtend.team35.yubiboard.camera.toCorrectedBitmap
import com.nxtend.team35.yubiboard.diagnostics.AppDiagnostics
import org.opencv.android.OpenCVLoader
import org.opencv.android.Utils
import org.opencv.core.Mat
import org.opencv.imgproc.Imgproc
import org.opencv.objdetect.ArucoDetector
import org.opencv.objdetect.Objdetect

class ArucoMarkerProcessor(
    private val onResult: (MarkerDetectionResult) -> Unit,
    private val onError: (Throwable) -> Unit,
) {
    private val stabilityTracker = MarkerStabilityTracker()
    private val detector: ArucoDetector?

    init {
        detector = runCatching {
            check(OpenCVLoader.initLocal()) { "OpenCV initialization failed" }
            ArucoDetector(Objdetect.getPredefinedDictionary(Objdetect.DICT_4X4_50))
        }.onFailure(onError).getOrNull()
    }

    fun process(image: ImageProxy, mirror: Boolean = false) {
        val activeDetector = detector
        if (activeDetector == null) {
            image.close()
            return
        }
        val capturedAt = SystemClock.uptimeMillis()
        runCatching {
            // ArUco画像を鏡像にすると辞書のビット列が変わり認識できないため、
            // 検出は元画像で行い、検出後の座標だけをプレビュー座標へ反転する。
            val bitmap = image.toCorrectedBitmap()
            val rgba = Mat()
            val gray = Mat()
            val ids = Mat()
            val corners = mutableListOf<Mat>()
            try {
                Utils.bitmapToMat(bitmap, rgba)
                Imgproc.cvtColor(rgba, gray, Imgproc.COLOR_RGBA2GRAY)
                activeDetector.detectMarkers(gray, corners, ids)
                val rawMarkers = corners.mapIndexedNotNull { index, cornerMat ->
                    val id = ids.get(index, 0)?.firstOrNull()?.toInt() ?: return@mapIndexedNotNull null
                    if (id !in MarkerStabilityTracker.EXPECTED_IDS) return@mapIndexedNotNull null
                    val normalizedCorners = (0 until 4).mapNotNull { cornerIndex ->
                        cornerMat.get(0, cornerIndex)?.let { point ->
                            NormalizedPoint(
                                x = (point[0] / bitmap.width).toFloat(),
                                y = (point[1] / bitmap.height).toFloat(),
                            )
                        }
                    }
                    if (normalizedCorners.size != 4) return@mapIndexedNotNull null
                    DetectedMarker(
                        id = id,
                        center = NormalizedPoint(
                            normalizedCorners.map { it.x }.average().toFloat(),
                            normalizedCorners.map { it.y }.average().toFloat(),
                        ),
                        corners = normalizedCorners,
                    )
                }.sortedBy { it.id }
                // ID 10=左上、11=右上というレイアウト検証は、鏡像化前のカメラ
                // 座標で行う。検証後に表示・送信用座標だけを水平反転する。
                val isStable = stabilityTracker.update(rawMarkers)
                val markers = if (mirror) {
                    rawMarkers.map(DetectedMarker::mirrorHorizontally)
                } else {
                    rawMarkers
                }
                AppDiagnostics.increment("aruco.frames")
                if (isStable) AppDiagnostics.increment("aruco.stable_frames")
                AppDiagnostics.gauge("aruco.marker_count", markers.size)
                AppDiagnostics.gauge("aruco.stable", isStable)
                AppDiagnostics.gauge(
                    "aruco.stability_progress",
                    "${stabilityTracker.stableFrameCount}/${MarkerStabilityTracker.DEFAULT_REQUIRED_STABLE_FRAMES}",
                )
                AppDiagnostics.sampled(
                    "aruco_result",
                    "vision",
                    "aruco_result",
                    mapOf(
                        "ids" to markers.joinToString { it.id.toString() },
                        "stable" to isStable,
                        "stableFrames" to stabilityTracker.stableFrameCount,
                        "mirrored" to mirror,
                        "centers" to markers.joinToString(";") { "${it.id}:${it.center.x},${it.center.y}" },
                    ),
                )
                onResult(
                    MarkerDetectionResult(
                        capturedAtMonotonicMs = capturedAt,
                        sourceWidth = bitmap.width,
                        sourceHeight = bitmap.height,
                        markers = markers,
                        stable = isStable,
                        stableFrameCount = stabilityTracker.stableFrameCount,
                        requiredStableFrames = MarkerStabilityTracker.DEFAULT_REQUIRED_STABLE_FRAMES,
                        cameraFacing = if (mirror) "front" else "back",
                    ),
                )
            } finally {
                corners.forEach(Mat::release)
                ids.release()
                gray.release()
                rgba.release()
            }
        }.onFailure {
            stabilityTracker.reset()
            AppDiagnostics.event("vision", "aruco_error", mapOf("message" to it.message))
            onError(it)
        }
    }

    fun reset() {
        stabilityTracker.reset()
        AppDiagnostics.event("vision", "aruco_reset")
    }
}

File: ios/Screact/Vision/ArucoMarkerProcessor.swift
import CoreVideo
import Foundation

/// Port of Android ArucoMarkerProcessor. Delegates raw detection to the OpenCV ObjC++ bridge
/// (DICT_4X4_50), keeps only the four expected IDs (10/11/12/13), normalizes corners, and confirms
/// stability via MarkerStabilityTracker — identical dictionary, IDs, and stability rules to Android.
final class ArucoMarkerProcessor {
    private let stabilityTracker = MarkerStabilityTracker()
    private let bridge = ArucoBridge()
    private let onResult: (MarkerDetectionResult) -> Void
    private let onError: (Error) -> Void

    /// True once OpenCV is linked (after `pod install`). When false, detection is a no-op.
    var isAvailable: Bool { ArucoBridge.isAvailable }

    init(onResult: @escaping (MarkerDetectionResult) -> Void,
         onError: @escaping (Error) -> Void) {
        self.onResult = onResult
        self.onError = onError
        if !ArucoBridge.isAvailable {
            onError(NSError(domain: "Screact.Aruco", code: 1,
                            userInfo: [NSLocalizedDescriptionKey:
                                "OpenCV未リンク。`pod install` 後に ArUco 検出が有効になります。"]))
        }
    }

    func process(_ pixelBuffer: CVPixelBuffer, capturedAtMonotonicMs: Int64) {
        guard ArucoBridge.isAvailable else { return }
        let raw = bridge.detectMarkers(in: pixelBuffer)
        var width = CVPixelBufferGetWidth(pixelBuffer)
        var height = CVPixelBufferGetHeight(pixelBuffer)

        var markers: [DetectedMarker] = []
        for entry in raw {
            guard let id = entry["id"] as? Int,
                  MarkerStabilityTracker.expectedIds.contains(id),
                  let flat = entry["corners"] as? [NSNumber], flat.count == 8 else { continue }
            if let w = entry["width"] as? Int { width = w }
            if let h = entry["height"] as? Int { height = h }
            let w = Float(max(width, 1))
            let h = Float(max(height, 1))
            var corners: [NormalizedPoint] = []
            for i in stride(from: 0, to: 8, by: 2) {
                corners.append(NormalizedPoint(x: flat[i].floatValue / w, y: flat[i + 1].floatValue / h))
            }
            let cx = corners.map { $0.x }.reduce(0, +) / Float(corners.count)
            let cy = corners.map { $0.y }.reduce(0, +) / Float(corners.count)
            markers.append(DetectedMarker(id: id, center: NormalizedPoint(x: cx, y: cy), corners: corners))
        }
        markers.sort { $0.id < $1.id }

        let isStable = stabilityTracker.update(markers)
        AppDiagnostics.shared.increment("aruco.frames")
        if isStable { AppDiagnostics.shared.increment("aruco.stable_frames") }
        AppDiagnostics.shared.gauge("aruco.marker_count", markers.count)
        AppDiagnostics.shared.gauge("aruco.stable", isStable)

        onResult(MarkerDetectionResult(
            capturedAtMonotonicMs: capturedAtMonotonicMs,
            sourceWidth: width,
            sourceHeight: height,
            markers: markers,
            stable: isStable,
            stableFrameCount: stabilityTracker.stableFrameCount,
            requiredStableFrames: MarkerStabilityTracker.defaultRequiredStableFrames
        ))
    }

    func reset() {
        stabilityTracker.reset()
        AppDiagnostics.shared.event("vision", "aruco_reset")
    }
}
